electrum/wallet.py

```diff
@@ -1960,22 +1988,39 @@ def fee_estimator(size: Union[int, float, Decimal]) -> int:
             #       forever. see #5433
             # note: Actually, it might be the case that not all UTXOs from the wallet are
             #       being spent if the user manually selected UTXOs.
-            sendable = sum(map(lambda c: c.value_sats(), coins))
-            for (_,i) in i_max:
-                outputs[i].value = 0
+            def distribute_amount(amount):
+                if amount < 0:
+                    raise NotEnoughFunds()
+                distr_amount = 0
+                for (weight, i) in i_max:
+                    # fixme: this does not check that value >= dust_threshold
+                    val = int((amount/i_max_sum) * weight)
+                    outputs[i].value = val
+                    distr_amount += val
+                (x,i) = i_max[-1]
+                outputs[i].value += (amount - distr_amount)
+
+            distribute_amount(0)
             tx = PartialTransaction.from_io(list(coins), list(outputs))
             fee = fee_estimator(tx.estimated_size())
-            amount = sendable - tx.output_value() - fee
-            if amount < 0:
-                raise NotEnoughFunds()
-            distr_amount = 0
-            for (weight, i) in i_max:
-                val = int((amount/i_max_sum) * weight)
-                outputs[i].value = val
-                distr_amount += val
-
-            (x,i) = i_max[-1]
-            outputs[i].value += (amount - distr_amount)
+
+            input_amount = sum(c.value_sats() for c in coins)
+            allocated_amount = sum(o.value for o in outputs if not parse_max_spend(o.value))
+            to_distribute = input_amount - allocated_amount - fee
+            distribute_amount(to_distribute)
+
+            amount_retained = sum(o.value for o in outputs if self.is_mine(o.address))
+            to_be_spent = input_amount - amount_retained # this includes the fee
+
+            if self.should_keep_reserve_utxo(to_be_spent, is_anchor_channel_opening):
+                self.logger.info(f'Adding change output to meet utxo reserve requirements')
+                change_addr = self.get_change_addresses_for_new_transaction(change_addr)[0]
+                change = PartialTxOutput.from_address_and_value(change_addr, self.config.LN_UTXO_RESERVE)
+                change.is_utxo_reserve = True # for GUI, using hasattr()
```

Comment:
why using hasattr? would it not be cleaner to just declare it in transaction.py?

```diff
diff --git a/electrum/transaction.py b/electrum/transaction.py
index 3abc4879db..78ab6c35ab 100644
--- a/electrum/transaction.py
+++ b/electrum/transaction.py
@@ -1875,6 +1875,7 @@ class PartialTxOutput(TxOutput, PSBTSection):
         self._script_descriptor = None  # type: Optional[Descriptor]
         self.is_mine = False  # type: bool  # whether the wallet considers the output to be ismine
         self.is_change = False  # type: bool  # whether the wallet considers the output to be change
+        self.is_utxo_reserve = False  # type: bool  # whether the wallet added this as anchor reserve
 
     @property
     def pubkeys(self) -> Set[bytes]:

```

Reply:
done